agent/agent.go

```diff
@@ -2009,6 +2016,32 @@ func (a *Agent) AddCheck(check *structs.HealthCheck, chkType *structs.CheckType,
 			monitor.Start()
 			a.checkMonitors[check.CheckID] = monitor
 
+		case chkType.IsAlias():
+			if existing, ok := a.checkAliases[check.CheckID]; ok {
+				existing.Stop()
+				delete(a.checkAliases, check.CheckID)
+			}
+			if chkType.Interval < checks.MinInterval {
+				a.logger.Printf("[WARN] agent: check '%s' has interval below minimum of %v",
+					check.CheckID, checks.MinInterval)
+				chkType.Interval = checks.MinInterval
+			}
+
+			var rpcReq structs.NodeSpecificRequest
+			rpcReq.Datacenter = a.config.Datacenter
+			rpcReq.Token = a.tokens.AgentToken()
```

Comment:
Is this the correct behavior? Should we use the agent token if none is specified, or user token?

Reply:
Is this check pretty similar to a prepared query and for prepared queries we allow configuring in the query definition.

So maybe we should add a "Token" parameter to the check definition and use that if provided or the anonymous token otherwise.

Reply:
Checks actually already support a `token` used for anti-entropy for that check. That might satisfy the behavior @banks is expecting above. We've never passed that into a check implementation so it'd require a bit more plumbing (its not on the struct), but that might be worth it.

So proposed logic:

  1. If `token` is specified on check, use that.
  2. Else, user agent token if specified.
  3. Else, no token.

Thoughts?

Reply:
@mitchellh I _think_ the least surprising thing would be if we default to copying the current scope token (from node/service registration) into the Check's `token` if it's not other wise specified.

After that I agree with your proposed logic being sane. My worry is that having to specify the same token in the service deviation _and_ each check body within that definition seems like a bad UX even if it is only needed in an edge case where the check doesn't just get access it needs anonymously.

Reply:
I looked into this a bit further and this is the behavior with checks (with an asterisk... will note):

  * For checks embedded in a service registration, the token used is the token specified with the service registration (directly in the definition). If not specified, the agent token is used. 
  * **Note on the above:** If a token is explicitly specified _in the embedded check_, it is _ignored_. I don't see this documented anywhere and its unlikely anyone has tried this, but that's how it works.
  * For standalone checks, the token on the check definition is used, otherwise the agent token.

This aligns really nicely with what is proposed above and therefore is probably the right decision: 1.) we should use the registration token 2.) else agent token. That matches everything else. 

I won't change the behavior of ignoring the embedded check token for service registration, since that seems designed that way.

Reply:
https://github.com/hashicorp/consul/pull/4320/commits/d6ecd97d1dffa72a9e928e66d87889deb9c09c7f